sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java

```diff
@@ -288,7 +288,7 @@ public void setDecimal(int ordinal, Decimal value, int precision) {
       Platform.putLong(baseObject, baseOffset + cursor, 0L);
       Platform.putLong(baseObject, baseOffset + cursor + 8, 0L);
 
-      if (value == null) {
+      if (value == null || !value.changePrecision(precision, value.scale())) {
```

Comment:
Thanks to @allisonwang-db for catching this bug!

Reply:
This looks valid for branch-3.0 and branch-2.4. Do you think we need to backport to to branch-2.4?

Reply:
Yes we should